pkg/controller/miganalytic/miganalytics_controller.go

```diff
@@ -0,0 +1,589 @@
+
+// ReconcileMigAnalytic reconciles a MigAnalytic object
+type ReconcileMigAnalytic struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+func (r *ReconcileMigAnalytic) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	var err error
+	log.Reset()
+
+	// Fetch the MigAnalytic instance
+	analytic := &migapi.MigAnalytic{}
+
+	err = r.Get(context.TODO(), request.NamespacedName, analytic)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Report reconcile error.
+	defer func() {
+		if err == nil || errors.IsConflict(err) {
+			return
+		}
+		analytic.Status.SetReconcileFailed(err)
+		err := r.Update(context.TODO(), analytic)
+		if err != nil {
+			log.Trace(err)
+			return
+		}
+	}()
+
+	// Begin staging conditions.
+	analytic.Status.BeginStagingConditions()
+
+	// Add Labels and Annoations
+	analytic.ObjectMeta.Labels = map[string]string{"migplan": analytic.Spec.MigPlanRef.Name}
+	analytic.ObjectMeta.Annotations = map[string]string{"migplan": analytic.Spec.MigPlanRef.Name}
+
+	// Validations.
+	err = r.validate(analytic)
+	if err != nil {
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Update Validation Status
+	err = r.Update(context.TODO(), analytic)
+	if err != nil {
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Analyze
+	if !analytic.Status.HasBlockerCondition() {
+		err = r.analyze(analytic)
+	} else {
+		return reconcile.Result{RequeueAfter: time.Second * RequeueInterval}, nil
+	}
+
+	// Ready
+	analytic.Status.SetReady(
+		!analytic.Status.HasBlockerCondition(),
+		ReadyMessage)
+
+	// End staging conditions.
+	analytic.Status.EndStagingConditions()
+
+	// Apply changes.
+	analytic.MarkReconciled()
+	err = r.Update(context.TODO(), analytic)
+	if err != nil {
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Done
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileMigAnalytic) analyze(analytic *migapi.MigAnalytic) error {
+	plan := &migapi.MigPlan{}
+
+	err := r.getAnalyticPlan(analytic, plan)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	srcCluster, err := plan.GetSourceCluster(r)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	client, err := srcCluster.GetClient(r)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	discovery, err := discovery.NewDiscoveryClientForConfig(client.RestConfig())
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	resources, err := collectResources(discovery)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	dynamic, err := dynamic.NewForConfig(client.RestConfig())
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	analytic.Status.Analytics.Plan = plan.Name
+	log.Info("Beginning analysis of plan.", "migplan", plan.Name)
+
+	for i, namespace := range plan.Spec.Namespaces {
+		log.Info("Beginning analysis of namespace.", "namespace", namespace)
```

Comment:
@jmontleon I recommend we remove this log statement. We convey all needed info to users (except for unhandled errs) via `Status.Conditions`  in mig-controller.